src/FamilyHubs.Referral.Core/ConnectTelemetryPiiRedactor.cs

```diff
@@ -0,0 +1,199 @@
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+
+namespace FamilyHubs.SharedKernel.Telemetry;
+
+/// <summary>
+/// Redacts Personally Identifiable Information (PII) from telemetry data we send to App Insights.
+/// What's redacted:
+///     postcode: a postcode can map to a single residential address, so can be used to identify an individual
+///     latitude & longitude: can be used to map to the postcode
+///
+/// Notes
+/// * ExceptionTelemetry is left alone (no PII currently comes through ExceptionTelemetry).
+///   We have to ensure we don't add any PII to exceptions.
+/// * We don't redact the console log output as that doesn't get persisted anywhere.
+/// </summary>
+/// <remarks>
+/// See
+/// https://learn.microsoft.com/en-us/azure/azure-monitor/app/api-filtering-sampling
+/// https://ico.org.uk/for-organisations/guide-to-data-protection/guide-to-the-general-data-protection-regulation-gdpr/key-definitions/what-is-personal-data/
+/// </remarks>>
+public class ConnectTelemetryPiiRedactor : ITelemetryInitializer
+{
+    // longtitude is due to the spelling error in the API. at some point, we should fix that (and all the consumers)
+    private static readonly Regex SiteQueryStringRegex = new(@"(?<=(postcode|latitude|longitude|longtitude)=)[^&]+", RegexOptions.Compiled);
+    private static readonly Regex ApiQueryStringRegex = new(@"(?<=latitude=)([-+]?[0-9]*\.?[0-9]+)(?=&)|(?<=longitude=)([-+]?[0-9]*\.?[0-9]+)(?=&)");
+    private static readonly Regex PathRegex = new(@"(?<=postcodes\/)[\w% ]+", RegexOptions.Compiled);
+    private static readonly string[] TraceItemsToRedact = { "postcode", "latitude", "longitude" };
+
+
+    public void Initialize(ITelemetry telemetry)
+    {
+        switch (telemetry)
+        {
+            case DependencyTelemetry dependencyTelemetry:
+                if (dependencyTelemetry.Name is "GET /api/services")
+                {
+                    // command name is obsolete and has been replaced by Data, but should contain the same as Data
+#pragma warning disable CS0618
+                    dependencyTelemetry.CommandName = dependencyTelemetry.Data =
+                        Sanitize(ApiQueryStringRegex, dependencyTelemetry.Data);
+#pragma warning restore CS0618
+                }
+                else if (dependencyTelemetry.Name.StartsWith("GET /postcodes/"))
+                {
+#pragma warning disable CS0618
+                    dependencyTelemetry.CommandName =
+                        dependencyTelemetry.Data = Sanitize(PathRegex, dependencyTelemetry.Data);
+#pragma warning restore CS0618
+                    dependencyTelemetry.Name = Sanitize(PathRegex, dependencyTelemetry.Name);
+                    dependencyTelemetry.Data = Sanitize(ApiQueryStringRegex, dependencyTelemetry.Data);
+                }
+                break;
+            case TraceTelemetry traceTelemetry:
+                foreach (var item in TraceItemsToRedact)
```

Comment:
This code will get called a lot, so we have to be aware of the amount of work we're doing - we want the code as optimised as is reasonable.

As it stands, this code can Sanitize the same string repeatedly. Whilst it means we end up with correctly redacted telemetries, it also means we're doing a lot of unnecessary work.

Also, checking the value of every property, even where some will never contain PII, means that we're doing unnecessary work (and where we do need to sanitize a property, the property can end up being sanitized repeatedly).

Can you please update the code so that messages & properties are only sanitized once, and to be more selective in the properties that are checked.